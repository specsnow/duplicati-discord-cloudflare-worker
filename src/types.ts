interface BackendStatistics {
  RemoteCalls: number;
  BytesUploaded: number;
  BytesDownloaded: number;
  FilesUploaded: number;
  FilesDownloaded: number;
  FilesDeleted: number;
  FoldersCreated: number;
  RetryAttempts: number;
  UnknownFileSize: number;
  UnknownFileCount: number;
  KnownFileCount: number;
  KnownFileSize: number;
  LastBackupDate: string;
  BackupListCount: number;
  TotalQuotaSpace: number;
  FreeQuotaSpace: number;
  AssignedQuotaSpace: number;
  ReportedQuotaError: boolean;
  ReportedQuotaWarning: boolean;
  MainOperation: string;
  ParsedResult: ParsedResult;
  Interrupted: boolean;
  Version: string;
  EndTime: string;
  BeginTime: string;
  Duration: string;
  MessagesActualLength: number;
  WarningsActualLength: number;
  ErrorsActualLength: number;
}

interface CompactResults {
  DeletedFileCount: number;
  DownloadedFileCount: number;
  UploadedFileCount: number;
  DeletedFileSize: number;
  DownloadedFileSize: number;
  UploadedFileSize: number;
  Dryrun: boolean;
  VacuumResults: null;
  MainOperation: string;
  ParsedResult: ParsedResult;
  Interrupted: boolean;
  Version: string;
  EndTime: string;
  BeginTime: string;
  Duration: string;
  MessagesActualLength: number;
  WarningsActualLength: number;
  ErrorsActualLength: number;
  BackendStatistics: BackendStatistics;
}

interface Verification {
  Key: string;
  Value: any[];
}

interface TestResults {
  MainOperation: string;
  VerificationsActualLength: number;
  Verifications: Verification[];
  ParsedResult: ParsedResult;
  Interrupted: boolean;
  Version: string;
  EndTime: string;
  BeginTime: string;
  Duration: string;
  MessagesActualLength: number;
  WarningsActualLength: number;
  ErrorsActualLength: number;
  BackendStatistics: BackendStatistics;
}

interface DeleteResults {
  DeletedSetsActualLength: number;
  DeletedSets: any[];
  Dryrun: boolean;
  MainOperation: string;
  CompactResults: CompactResults;
  ParsedResult: ParsedResult;
  Interrupted: boolean;
  Version: string;
  EndTime: string;
  BeginTime: string;
  Duration: string;
  MessagesActualLength: number;
  WarningsActualLength: number;
  ErrorsActualLength: number;
  BackendStatistics: BackendStatistics;
}

interface Data {
  DeletedFiles: number;
  DeletedFolders: number;
  ModifiedFiles: number;
  ExaminedFiles: number;
  OpenedFiles: number;
  AddedFiles: number;
  SizeOfModifiedFiles: number;
  SizeOfAddedFiles: number;
  SizeOfExaminedFiles: number;
  SizeOfOpenedFiles: number;
  NotProcessedFiles: number;
  AddedFolders: number;
  TooLargeFiles: number;
  FilesWithError: number;
  ModifiedFolders: number;
  ModifiedSymlinks: number;
  AddedSymlinks: number;
  DeletedSymlinks: number;
  PartialBackup: boolean;
  Dryrun: boolean;
  MainOperation: string;
  CompactResults: CompactResults;
  VacuumResults: null;
  DeleteResults: DeleteResults;
  RepairResults: null;
  TestResults: TestResults;
  ParsedResult: ParsedResult;
  Interrupted: boolean;
  Version: string;
  EndTime: string;
  BeginTime: string;
  Duration: string;
  MessagesActualLength: number;
  WarningsActualLength: number;
  ErrorsActualLength: number;
  BackendStatistics: BackendStatistics;
}

interface Extra {
  OperationName: string;
  'machine-id': string;
  'machine-name': string;
  'backup-name': string;
  'backup-id': string;
}

export interface DuplicatiResponse {
  Data: Data;
  Extra: Extra;
  LogLines: any[];
  Exception: null;
}

export type ParsedResult = 'Success' | 'Warning' | 'Error' | 'Fatal';

interface Attachment {
  id: string
  filename: string
  size: number
  url: string
  proxy_url: string
  height?: number
  width: number
}

interface Author {
  name?: string
  url?: string
  icon_url?: string | Attachment
  proxy_icon_url?: string
}
interface Field {
  name: string
  value: string
  inline?: boolean
}

interface Footer {
  text: string
  icon_url?: string | Attachment
  proxy_icon_url?: string
}
interface Image {
  url: string | Attachment
  proxy_url?: string
  height?: number
  width?: number
}
interface Provider {
  name?: string
  url?: string
}
interface Thumbnail {
  url: string | Attachment
  proxy_url?: string
  height?: number
  width?: number
}

interface Video {
  url: string
  height?: number
  width?: number
}

export interface Embed {
  title?: string
  type?: 'rich'
  url?: string
  description?: string
  timestamp?: string
  color?: number
  footer?: Footer
  image?: Image
  thumbnail?: Thumbnail
  video?: Video
  provider?: Provider
  author?: Author
  fields?: Array<Field>
}