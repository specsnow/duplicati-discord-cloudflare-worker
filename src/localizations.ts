const locales = {
  enUS: {
    author: {
      name: 'Duplicati Discord Notification'
    },
    states: {
      started: 'Started',
      finished: 'Finished',
      interrupted: 'Interrupted',
    },
    boolean: {
      yes: 'Yes',
      no: 'No',
    },
    fields: {
      additionalInformation: 'Additional information',
      duration: 'Duration',
      files: 'Files',
      deletedFiles: 'Deleted Files',
      modifiedFiles: 'Modified Files',
      addedFiles: 'Added Files',
      size: 'Size',
    },
    footer: {
      operation: 'Operation',
      machineName: 'Machine Name',
    }
  }
} as const

type Locale = keyof typeof locales

type Paths<
  T extends Record<string, any>,
  P extends string = ''
> = {
  [K in keyof T]: T[K] extends Record<string, any>
    ? Paths<T[K], `${P}${P extends '' ? '' : '.'}${string & K}`>
    : `${P}${P extends '' ? '' : '.'}${string & K}`
}[keyof T]

type GetPaths = Paths<typeof locales.enUS>

/**
 * Function that accepts the locale and returns a function to get the localized string by dot notation.
 */
export function getTranslation(locale: Locale = 'enUS') {
  return (key: GetPaths): string => {
    const data = locales[locale]
    
    // @ts-expect-error
    return key.split('.').reduce((acc, current) => acc?.[current], data)
  }
}