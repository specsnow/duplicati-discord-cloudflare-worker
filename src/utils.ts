/**
 * Convert date-time string to human-readable format.
 * @example
 * formatDate('2025-03-23T22:28:59.0337138Z') // '2025-03-23 22:28:59'
 * @example
 * formatDate('2025-01-26T06:30:00+01:00') // '2025-01-26 06:30:00'
 */
export function formatDate(dateTime: string): string {
  return new Date(dateTime).toISOString().replace(/T/, ' ').replace(/\..+/, '') + ' UTC'
}

/**
 * Convert a duration to human-readable format.
 * @example
 * formatDuration('00:00:49.6596191') // '49.66 seconds'
 * @example
 * formatDuration('00:01:00.0000000') // '1 minute'
 * @example
 * formatDuration('00:01:30.0000000') // '1 minute 30 seconds'
 * @example
 * formatDuration('01:15:30.8948789') // '1 hour 15 minutes 30 seconds'
 */
export function formatDuration(duration: string): string {
  const [hours, minutes, secondsAndMilliseconds] = duration.split(':')
  const [seconds] = secondsAndMilliseconds.split('.')
  const parts: string[] = []

  if (hours !== '00') {
    parts.push(`${parseInt(hours)} hour${parseInt(hours) > 1 ? 's' : ''}`)
  }
  if (minutes !== '00') {
    parts.push(`${parseInt(minutes)} minute${parseInt(minutes) > 1 ? 's' : ''}`)
  }
  if (seconds !== '00') {
    parts.push(`${parseInt(seconds)} second${parseInt(seconds) > 1 ? 's' : ''}`)
  }

  return parts.join(' ')
}

/**
 * Convert bytes to human-readable format.
 * @example
 * prettyBytes(1024) // '1 KB'
 * @example
 * prettyBytes(7714963409) // '7.19 GB'
 */
export function prettyBytes(bytes: number): string {
  if (bytes === 0) return '0 B'
  const k = 1024
  const i = Math.floor(Math.log(bytes) / Math.log(k))
  const sizes = ['B', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB']

  return `${Number.parseFloat((bytes / k ** i).toFixed(2))} ${sizes[i]}`
}