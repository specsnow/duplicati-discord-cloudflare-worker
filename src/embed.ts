import { formatDate, formatDuration, prettyBytes } from './utils'
import { getTranslation } from './localizations'
import type { Embed, DuplicatiResponse, ParsedResult } from './types'

const t = getTranslation()

const ICONS: Record<ParsedResult, string> = {
  Success: ':white_check_mark:',
  Warning: ':warning:',
  Error: ':x:',
  Fatal: ':fire:'
}

const COLORS: Record<ParsedResult, number> = {
  Success: 0x00A53D,
  Warning: 0xF44900,
  Error: 0xE7000B,
  Fatal: 0xE7000B
}

interface DiscordEmbed {
  embeds: Embed[]
}

/**
 * Return an embed for an unhandled error.
 */
function unhandledError({ name, exception, reason }: { name?: string; exception?: string; reason?: string }): DiscordEmbed {
  return {
    embeds: [
      {
        author: {
          name: t('author.name'),
        },
        title: `${ICONS.Fatal} ${name ? `${name}: ` : ''}Unhandled Error`,
        description: `${exception ? `**Exception**: ${exception}` : ''}${exception && reason ? '\n' : ''}${reason ? `**Reason**: ${reason}` : ''}`,
        color: COLORS.Fatal,
      }
    ]
  }
}

export function createEmbed(res: DuplicatiResponse): DiscordEmbed {
  if (!res) {
    return unhandledError({ reason: 'No response from Duplicati' })
  }
  if (res.Exception) {
    return unhandledError({ exception: res.Exception })
  }
  if (!res.Data) {
    return unhandledError({ reason: 'No data in response' })
  }

  const { Data, Extra } = res
  const { 'backup-name': backupName, OperationName: operationName, 'machine-name': machineName } = Extra

  return {
    embeds: [
      {
        author: {
          name: t('author.name'),
        },
        title: `${ICONS[Data.ParsedResult]} ${backupName}`,
        color: COLORS[Data.ParsedResult],
        footer: {
          text: `${t('footer.operation')}: ${operationName} | ${t('footer.machineName')}: ${machineName}`
        },
        fields: [
          {
            name: t('fields.duration'),
            value: formatDuration(Data.Duration),
            inline: true
          },
          {
            name: t('fields.files'),
            value: Data.ExaminedFiles.toString(),
            inline: true
          },
          {
            name: t('fields.size'),
            value: prettyBytes(Data.SizeOfExaminedFiles),
            inline: true
          },
          {
            name: t('fields.addedFiles'),
            value: Data.AddedFiles.toString(),
            inline: true
          },
          {
            name: t('fields.modifiedFiles'),
            value: Data.ModifiedFiles.toString(),
            inline: true
          },
          {
            name: t('fields.deletedFiles'),
            value: Data.DeletedFiles.toString(),
            inline: true
          },
          {
            name: t('fields.additionalInformation'),
            value: `${t('states.started')}: ${formatDate(Data.BeginTime)}\n${t('states.finished')}: ${formatDate(Data.EndTime)}\n${t('states.interrupted')}: ${Data.Interrupted ? t('boolean.yes') : t('boolean.no')}`
          }
        ]
      }
    ]
  }
}