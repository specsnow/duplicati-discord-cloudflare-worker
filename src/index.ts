import { Hono } from 'hono'
import { html } from 'hono/html'
import { validator } from 'hono/validator'
import type { DuplicatiResponse } from './types'
import { createEmbed } from './embed'

const app = new Hono()

app.get('/', (c) => {
  return c.html(html`<!doctype html><body><p>Please use the instructions at <a href="https://github.com/LekoArts/duplicati-discord-cloudflare-worker">duplicati-discord-cloudflare-worker</a> to self-host this worker.</a></body>`)
})

app.post('/:channel_id/:webhook_id', validator('json', (value) => value), async (c) => {
  const { channel_id, webhook_id } = c.req.param()
  const thread_id = c.req.query('thread_id')
  let discordWebhookUrl = `https://discord.com/api/webhooks/${channel_id}/${webhook_id}`

  if (thread_id) {
    // See https://discord.com/developers/docs/resources/webhook#execute-webhook
    discordWebhookUrl += `?thread_id=${encodeURIComponent(thread_id)}`
  }

  const body = c.req.valid('json') as DuplicatiResponse
  const discordEmbed = createEmbed(body)

  const res = await fetch(discordWebhookUrl, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(discordEmbed)
  })

  if (!res.ok) {
    return c.json({ status: res.status , error: res.statusText })
  }

  return c.json({ message: 'Discord webhook successfully triggered' }, 200)
})

export default app
