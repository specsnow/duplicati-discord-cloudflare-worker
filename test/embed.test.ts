import { describe, expect, it } from 'vitest'
import { createEmbed } from '../src/embed'
import { success, warning } from './fixtures/duplicati'

describe('createEmbed', () => {
  it('should return an embed for a valid Duplicati response', () => {
    // @ts-expect-error
    expect(createEmbed(success)).toMatchSnapshot()
  })
  it('should return an embed for a warning', () => {
    // @ts-expect-error
    expect(createEmbed(warning)).toMatchSnapshot()
  })
  it('should return an embed for empty input', () => {
    // @ts-expect-error
    expect(createEmbed()).toMatchSnapshot()
  })
  it('should return an embed for an exception', () => {
    // @ts-expect-error
    expect(createEmbed({ Exception: 'Exception' })).toMatchSnapshot()
  })
  it('should return an embed for an empty Data', () => {
    // @ts-expect-error
    expect(createEmbed({ })).toMatchSnapshot()
  })
})
