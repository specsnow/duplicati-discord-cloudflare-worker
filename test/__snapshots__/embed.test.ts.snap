// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createEmbed > should return an embed for a valid Duplicati response 1`] = `
{
  "embeds": [
    {
      "author": {
        "name": "Duplicati Discord Notification",
      },
      "color": 42301,
      "fields": [
        {
          "inline": true,
          "name": "Duration",
          "value": "49 seconds",
        },
        {
          "inline": true,
          "name": "Files",
          "value": "504",
        },
        {
          "inline": true,
          "name": "Size",
          "value": "7.19 GB",
        },
        {
          "inline": true,
          "name": "Added Files",
          "value": "0",
        },
        {
          "inline": true,
          "name": "Modified Files",
          "value": "0",
        },
        {
          "inline": true,
          "name": "Deleted Files",
          "value": "0",
        },
        {
          "name": "Additional information",
          "value": "Started: 2025-03-23 22:28:09 UTC
Finished: 2025-03-23 22:28:59 UTC
Interrupted: No",
        },
      ],
      "footer": {
        "text": "Operation: Backup | Machine Name: test-machine-name",
      },
      "title": ":white_check_mark: Test Backup Name",
    },
  ],
}
`;

exports[`createEmbed > should return an embed for a warning 1`] = `
{
  "embeds": [
    {
      "author": {
        "name": "Duplicati Discord Notification",
      },
      "color": 16009472,
      "fields": [
        {
          "inline": true,
          "name": "Duration",
          "value": "49 seconds",
        },
        {
          "inline": true,
          "name": "Files",
          "value": "204",
        },
        {
          "inline": true,
          "name": "Size",
          "value": "7.19 GB",
        },
        {
          "inline": true,
          "name": "Added Files",
          "value": "0",
        },
        {
          "inline": true,
          "name": "Modified Files",
          "value": "0",
        },
        {
          "inline": true,
          "name": "Deleted Files",
          "value": "0",
        },
        {
          "name": "Additional information",
          "value": "Started: 2025-03-23 22:28:09 UTC
Finished: 2025-03-23 22:28:59 UTC
Interrupted: No",
        },
      ],
      "footer": {
        "text": "Operation: Backup | Machine Name: test-machine-name",
      },
      "title": ":warning: Test Backup Name",
    },
  ],
}
`;

exports[`createEmbed > should return an embed for an empty Data 1`] = `
{
  "embeds": [
    {
      "author": {
        "name": "Duplicati Discord Notification",
      },
      "color": 15138827,
      "description": "**Reason**: No data in response",
      "title": ":fire: Unhandled Error",
    },
  ],
}
`;

exports[`createEmbed > should return an embed for an exception 1`] = `
{
  "embeds": [
    {
      "author": {
        "name": "Duplicati Discord Notification",
      },
      "color": 15138827,
      "description": "**Exception**: Exception",
      "title": ":fire: Unhandled Error",
    },
  ],
}
`;

exports[`createEmbed > should return an embed for empty input 1`] = `
{
  "embeds": [
    {
      "author": {
        "name": "Duplicati Discord Notification",
      },
      "color": 15138827,
      "description": "**Reason**: No response from Duplicati",
      "title": ":fire: Unhandled Error",
    },
  ],
}
`;
